mino/minows/session.go

```diff
@@ -2,85 +2,186 @@ package minows
 
 import (
 	"context"
-	"encoding/json"
-	"errors"
-	"github.com/libp2p/go-libp2p/core/network"
+	"encoding/gob"
 	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
 	"go.dedis.ch/dela/mino"
 	"go.dedis.ch/dela/serde"
 	"golang.org/x/xerrors"
+	"io"
+	"sync"
 )
 
-type envelope struct {
-	author mino.Address
-	msg    serde.Message
-	err    error
+var ErrWrongAddressType = xerrors.New("wrong address type")
+var ErrNotPlayer = xerrors.New("not player")
+
+// MaxUnreadAllowed Maximum number of unread messages allowed
+// in orchestrator's incoming message buffer before pausing relaying
+const MaxUnreadAllowed = 1e3
+
+type Forward struct {
+	Packet
+	Destination []byte
 }
 
-// session represents a stream session started by rpc.Stream()
-// between a minows instance and other players of the RPC.
-// A session ends for all participants when the initiator is done and cancels
-// the stream context.
-// - implements mino.Sender, mino.Receiver
-type session struct {
-	rpc  rpc
-	in   chan envelope
-	outs map[peer.ID]*json.Encoder
+type orchestrator struct {
+	logger zerolog.Logger
+
+	myAddr orchestratorAddr
+	rpc    rpc
+	// Connects to the participants
+	outs map[peer.ID]*gob.Encoder
+	in   chan Packet
 }
 
-// Send sends a message to all addresses concurrently.
-// Send is asynchronous and returns immediately an error channel that
-// closes when the message has either been sent or errored to each address.
-func (s session) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+func (o orchestrator) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
 	send := func(addr mino.Address) error {
```

Comment:
- Make this a method of `orchestrator` - else it's capturing `o` here, which is confusing. If it's part of `orchestrator`, it's clear that it will use the fields of `orchestrator`
- Take `msg` as parameter, else the captured `msg` will be used both in this closure and be passed to `doSend`

Reply:
changed